execution_engine/src/storage/global_state/scratch.rs

```diff
@@ -0,0 +1,564 @@
+            let mut tmp = AdditiveMap::new();
+            for TestPair { key, value } in &test_pairs_updated {
+                if let StoredValue::CLValue(cl_value) = value {
+                    match cl_value.cl_type() {
+                        CLType::I32 => {
+                            tmp.insert(*key, Transform::AddInt32(2i32));
+                        }
+                        CLType::String => {}
+                        _ => unreachable!(
+                            "remember to update the test pairs if new cl_types are added"
+                        ),
+                    }
+                }
+            }
+            tmp
+        };
+        scratch
+            .commit(correlation_id, root_hash, add_effects.clone())
+            .unwrap();
+        let updated_hash = state2
+            .commit(correlation_id, updated_hash, add_effects)
+            .unwrap();
+
+        let scratch_checkout = scratch.checkout(root_hash).unwrap().unwrap();
+        let updated_checkout = state2.checkout(updated_hash).unwrap().unwrap();
+        let all_keys = updated_checkout
+            .keys_with_prefix(correlation_id, &[])
+            .unwrap();
+
+        // Check that cache matches the contents of the second instance of lmdb
+        for key in all_keys {
+            assert_eq!(
+                scratch_checkout
+                    .read(correlation_id, &key)
+                    .unwrap()
+                    .as_ref(),
+                updated_checkout
+                    .read(correlation_id, &key)
+                    .unwrap()
+                    .as_ref()
+            );
+        }
+    }
+
+    #[test]
+    fn commit_updates_state_and_original_state_stays_intact() {
+        let correlation_id = CorrelationId::new();
+        let test_pairs_updated = create_test_pairs_updated();
+
+        let TestState {
+            state, root_hash, ..
+        } = create_test_state();
+
+        let scratch = state.create_scratch();
+
+        let effects: AdditiveMap<Key, Transform> = {
+            let mut tmp = AdditiveMap::new();
+            for TestPair { key, value } in &test_pairs_updated {
+                tmp.insert(*key, Transform::Write(value.to_owned()));
+            }
+            tmp
+        };
+
+        let updated_hash = scratch.commit(correlation_id, root_hash, effects).unwrap();
+
+        let updated_checkout = scratch.checkout(updated_hash).unwrap().unwrap();
+        for TestPair { key, value } in test_pairs_updated.iter().cloned() {
+            assert_eq!(
+                Some(value),
+                updated_checkout.read(correlation_id, &key).unwrap()
+            );
+        }
+
+        let original_checkout = state.checkout(root_hash).unwrap().unwrap();
+        for TestPair { key, value } in create_test_pairs().iter().cloned() {
+            assert_eq!(
+                Some(value),
+                original_checkout.read(correlation_id, &key).unwrap()
```

Comment:
Can you add an assertion message for the error case? IIUC, it would fail if the original state gets update via `scratch` manipulation.

Reply:
GitHub still doesn't show me the error message in this `assert_eq!`.